{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "import numpy as np\n",
    "import torch\n",
    "import anndata as ad\n",
    "import networkx as nx\n",
    "import scanpy as sc\n",
    "import scglue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def set_seed(seed):\n",
    "    random.seed(seed)  \n",
    "    np.random.seed(seed)  \n",
    "    torch.manual_seed(seed)  \n",
    "    torch.cuda.manual_seed(seed) \n",
    "    torch.cuda.manual_seed_all(seed)  \n",
    "    torch.backends.cudnn.deterministic = True  \n",
    "    torch.backends.cudnn.benchmark = False  \n",
    "\n",
    "set_seed(42)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rna = sc.read(\"./data/neurips-multiome/rna_hvg.h5ad\")\n",
    "atac = sc.read(\"./data/neurips-multiome/atac_hvf.h5ad\")\n",
    "\n",
    "rna.X = rna.layers['counts'].copy()\n",
    "atac.X = atac.layers['counts'].copy()\n",
    "\n",
    "sc.pp.highly_variable_genes(rna, n_top_genes=2000, flavor=\"seurat_v3\")\n",
    "sc.pp.normalize_total(rna)\n",
    "sc.pp.log1p(rna)\n",
    "sc.pp.scale(rna)\n",
    "sc.tl.pca(rna, n_comps=100, svd_solver=\"auto\")\n",
    "rna.X = rna.layers['counts'].copy()\n",
    "atac.X = atac.layers['counts'].copy()\n",
    "\n",
    "scglue.data.lsi(atac, n_components=100, n_iter=15)\n",
    "\n",
    "scglue.data.get_gene_annotation(\n",
    "    rna,\n",
    "    gtf=\"./data/genes.gtf\",\n",
    "    gtf_by=\"gene_name\"\n",
    ")\n",
    "rna.var.loc[:, [\"chrom\", \"chromStart\", \"chromEnd\"]].head()\n",
    "\n",
    "split = atac.var_names.str.split(r\"[:-]\")\n",
    "atac.var[\"chrom\"] = split.map(lambda x: x[0])\n",
    "atac.var[\"chromStart\"] = split.map(lambda x: x[1]).astype(int)\n",
    "atac.var[\"chromEnd\"] = split.map(lambda x: x[2]).astype(int)\n",
    "\n",
    "rna.var = rna.var.loc[:, ~rna.var.columns.duplicated()]\n",
    "atac.var = atac.var.loc[:, ~atac.var.columns.duplicated()]\n",
    "\n",
    "for col in rna.var.columns:\n",
    "    if rna.var[col].dtype.name in [\"object\", \"category\"] and col != \"highly_variable\":\n",
    "        rna.var[col] = rna.var[col].astype(str)\n",
    "\n",
    "for col in atac.var.columns:\n",
    "    if atac.var[col].dtype.name in [\"object\", \"category\"]:\n",
    "        atac.var[col] = atac.var[col].astype(str)\n",
    "\n",
    "guidance = scglue.genomics.rna_anchored_guidance_graph(rna, atac)\n",
    "\n",
    "scglue.graph.check_graph(guidance, [rna, atac])\n",
    "\n",
    "scglue.models.configure_dataset(rna, \"NB\", use_highly_variable=True,use_layer=\"counts\", use_rep=\"X_pca\", use_batch='batch')\n",
    "\n",
    "scglue.models.configure_dataset(atac, \"NB\", use_highly_variable=True,use_rep=\"X_lsi\", use_batch='batch')\n",
    "\n",
    "glue = scglue.models.fit_SCGLUE({\"rna\": rna, \"atac\": atac}, guidance, model=scglue.models.PairedSCGLUEModel, fit_kws={\"directory\": \"glue\"})\n",
    "\n",
    "rna.obsm[\"latent\"] = glue.encode_data(\"rna\", rna)\n",
    "rna.write_h5ad(\"./results/neurips-multiome-gluer.h5ad\")\n",
    "\n",
    "atac.obsm[\"latent\"] = glue.encode_data(\"atac\", atac)\n",
    "atac.write_h5ad(\"./results/neurips-multiome-gluea.h5ad\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "multisi",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.23"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
